js/impress.js

```diff
@@ -659,6 +705,64 @@
         return lib;
     };
 
+    // `addPreInitPlugin` allows plugins to register a function that should
+    // be run (synchronously) at the beginning of init, before
+    // impress().init() itself executes.
+    impress.addPreInitPlugin = function( plugin, weight ) {
```

Comment:
Why is weight necessary? Are there plugins who rely on the order of execution?

Reply:
As a simple example, plugin-b might depend on the output (e.g. some DOM modifications) of plugin-a.

If you look at all the plugins in my repo, there's also other "logical" reasons to use some ordering. For example, substep plugin will prevent transitioning to next step, and instead reveals a substep on the same step. So it makes sense for that one to execute early, because it will in any case cancel the stepleave event.

With some other plugins it could be ambiguous/conflicting what should happen if user tries to use both, such as using both skip (skip next slide) and goto (jump to arbitrary slide). So the weight makes their priority explicit.

Reply:
> For example, substep plugin will prevent transitioning to next step, and instead reveals a substep on the same step. So it makes sense for that one to execute early, because it will in any case cancel the stepleave event.

In the case of the substep plugin, instead of the plugin author to decide the priority of the execution, couldn't the user decide it? If you want to intercept the event you would put the plugin in a `script` tag further down the list and the plugin would do something like `event.stopPropagation()`:

```html
<script src="impress.js"></script>
<script src="goto-plugin.js"></script>
<script src="substep-plugin.js"></script> <!-- This has priority on the event related to navigation, because it will prevent the default behavior from the "goto" plugin -->
```

**impress.js**
```javascript
.addEventListener( "impress:stepleave", function( event ) {
  // Impress handlers
});
```

**goto-plugin.js**
```javascript
.addEventListener( "impress:stepleave", function( event ) {
  // goto-plugin handlers
});
```

**substep-plugin.js**
```javascript
.addEventListener( "impress:stepleave", function( event ) {
  event.stopPropagation();
  // stepleave handlers
});
```

Reply:
I'm also not a big fun of this `weight` parameter, as it may be quite confusing for end users.
Luckily - this is quite 'internal' API used by plugin developers only now.

Yes, it would be great to have plugins being independent, but it's a theoretical ideal state that in practice would be very hard to avoid. Impress is a library working in browser DOM and modifying it, and plugins will rely and modify DOM, so there will be side-effects by definition and therefore plugins may conflict in some areas.

Moving freedom (and responsibility) of defining plugins order to impress.js user could be a 'solution' to this, but on the other hand it makes it harder for them to use plugins, because they would need to understand what plugins do internally. And really it's a plugin developer who knows best how it works and in what order it should be called. Yes, it has a downside of not scaling very well if external people would start to work on plugins, but honestly - we are really far from that and hopefully not many plugins would really need a control over when they need to be called (apart from internally developed plugins), so external developed plugins would be able to ignore weight altogether.

Reply:
The main point about weight is indeed that this is internal api to be used by plugins. Plugin developers will know how their plugin interacts with other plugins, if at all. There is no reason for presentation authors to be aware of it.

Because it is "internal", it can also be changed later if it's not sufficient. FWIW, Drupal plugins have a similar weight mechanism. So while it's simple, I believe it is scalable.